docs/define.rst

```diff
@@ -0,0 +1,340 @@
+
+Also see the `hookwrapper`_ section in the ``pytest`` docs.
+
+.. _specs:
+
+Specifications
+--------------
+A hook *specification* (*hookspec*) is a definition used to validate each
+*hookimpl* ensuring that an extension writer has correctly defined their
+callback function *implementation* .
+
+*hookspecs* are defined using similarly marked functions however only the
+function *signature* (its name and names of all its arguments) is analyzed
+and stored. As such, often you will see a *hookspec* defined with only
+a docstring in its body.
+
+*hookspecs* are loaded using the
+:py:meth:`~pluggy.PluginManager.add_hookspecs()` method and normally
+should be added before registering corresponding *hookimpls*:
+
+.. code-block:: python
+
+    import sys
+    from pluggy import PluginManager, HookspecMarker
+
+    hookspec = HookspecMarker('myproject')
+
+    @hookspec
+    def setup_project(config, args):
+        """This hook is used to process the inital config and input
+        arguments.
+        """
+
+    pm = PluginManager('myproject')
+
+    # load from the local module's namespace
+    pm.add_hookspecs(sys.modules[__name__])
+
+
+Registering a *hookimpl* which does not meet the constraints of its
+corresponding *hookspec* will result in an error.
+
+A *hookspec* can also be added **after** some *hookimpls* have been
+registered however this is not normally recommended as it results in
+delayed hook validation.
+
+.. note::
+    The term *hookspec* can sometimes refer to the plugin-namespace
+    which defines ``hookspec`` decorated functions as in the case of
+    ``pytest``'s `hookspec module`_
+
+Enforcing spec validation
+^^^^^^^^^^^^^^^^^^^^^^^^^
+By default there is no strict requirement that each *hookimpl* has
+a corresponding *hookspec*. However, if you'd like you enforce this
+behavior you can run a check with the
+:py:meth:`~pluggy.PluginManager.check_pending()` method. If you'd like
+to enforce requisite *hookspecs* but with certain exceptions for some hooks
+then make sure to mark those hooks as :ref:`optional <optionalhook>`.
+
+Opt-in arguments
+^^^^^^^^^^^^^^^^
+To allow for *hookspecs* to evolve over the lifetime of a project,
+*hookimpls* can accept **less** arguments then defined in the spec.
+This allows for extending hook arguments (and thus semantics) without
+breaking existing *hookimpls*.
+
+In other words this is ok:
+
+.. code-block:: python
+
+    @hookspec
+    def myhook(config, args):
+        pass
+
+    @hookimpl
+    def myhook(args):
+        print(args)
+
+
+whereas this is not:
+
+.. code-block:: python
+
+    @hookspec
+    def myhook(config, args):
+        pass
+
+    @hookimpl
+    def myhook(config, args, extra_arg):
+        print(args)
+
+.. _firstresult:
+
+First result only
+^^^^^^^^^^^^^^^^^
+A *hookspec* can be marked such that when the *hook* is called the call loop
+will only invoke up to the first *hookimpl* which returns a none ``None`` result.
```

Comment:
Oops found another "none none" occurrence, for consistency we probably need to rephrase this as well.